easybuild/scripts/findPythonDeps.py

```diff
@@ -0,0 +1,169 @@
+        if not matching_entries:
+            raise RuntimeError("Found no installed package for '%s' in %s" % (pkg, dep_tree))
+        if len(matching_entries) > 1:
+            raise RuntimeError("Found multiple installed packages for '%s' in %s" % (pkg, dep_tree))
+        entry = matching_entries[0]
+        res.append((entry['package']['package_name'], entry['package']['installed_version']))
+        deps = (dep['package_name'] for dep in entry['dependencies'])
+        res.extend(find_deps(deps, dep_tree))
+    return res
+
+
+def print_deps(package, verbose):
+    if verbose:
+        print('Getting dep tree of ' + package)
+    dep_tree = get_dep_tree(package, verbose)
+    if verbose:
+        print('Extracting dependencies of ' + package)
+    deps = find_deps([extract_pkg_name(package)], dep_tree)
+
+    installed_modules = {mod.project_name for mod in pkg_resources.working_set}
+    if verbose:
+        print("Installed modules: %s" % installed_modules)
+
+    # iterate over deps in reverse order, get rid of duplicates along the way
+    # also filter out Python packages that are already installed in current environment
+    res = []
+    handled = set()
+    for dep in reversed(deps):
+        if dep not in handled:
+            handled.add(dep)
+            if dep[0] in installed_modules:
+                if verbose:
+                    print("Skipping installed module '%s'" % dep[0])
+            else:
+                res.append(dep)
+
+    print("List of dependencies in (likely) install order:")
+    pprint(res, indent=4)
+    print("Sorted list of dependencies:")
+    pprint(sorted(res), indent=4)
+
+
+parser = argparse.ArgumentParser(
+    description='Find dependencies of Python packages by installing it in a temporary virtualenv. ',
+    epilog=' && '.join(['Example usage with EasyBuild: '
+                        'eb TensorFlow-2.3.4.eb --dump-env',
+                        'source TensorFlow-2.3.4.env',
+                        sys.argv[0] + ' tensorflow==2.3.4'])
+)
+parser.add_argument('package', metavar='python-pkg-spec',
+                    help='Python package spec, e.g. tensorflow==2.3.4')
+parser.add_argument('--ec', metavar='easyconfig', help='EasyConfig to use as the build environment')
+parser.add_argument('--verbose', help='Verbose output', action='store_true')
+args = parser.parse_args()
+
+if args.ec:
+    if not can_run('eb', '--version'):
+        print('EasyBuild not found or executable. Make sure it is in your $PATH when using --ec!')
+        sys.exit(1)
+    with temporary_directory() as tmp_dir:
+        old_dir = os.getcwd()
+        os.chdir(tmp_dir)
+        if args.verbose:
+            print('Running EasyBuild to get build environment')
+        run_cmd(['eb', args.ec, '--dump-env', '--force'], action_desc='Dump build environment')
+        os.chdir(old_dir)
+
+        cmd = 'source %s/*.env && %s "%s"' % (tmp_dir, sys.argv[0], args.package)
```

Comment:
Sourcing this makes the assumption that all the modules are actually installed, if they are not available you get (uncaught) Lmod errors like
```
Lmod has detected the following error: The following module(s) are unknown: "snappy/1.1.8-GCCcore-10.3.0"

Please check the spelling or version number. Also try "module spider ..."
It is also possible your cache file is out-of-date; it may help to try:
  $ module --ignore_cache load "snappy/1.1.8-GCCcore-10.3.0"

Also make sure that all modulefiles written in TCL start with the string #%Module
```
I don't know how much you can do about this. The implication is basically that you need (at least) the dependencies resolved of the starting easyconfig.

Reply:
Added to help, yes can't do much else

Reply:
Actually you can check for this:
```
[ocaisa@gpunode1 ~]$ eb TensorFlow-2.5.0-foss-2020b.eb --missing | grep '^*' | grep -v TensorFlow-2.5.0-foss-2020b.eb|wc -l
58
[ocaisa@gpunode1 ~]$ eb TensorFlow-2.3.1-foss-2020a-Python-3.8.2.eb --missing | grep '^*' | grep -v TensorFlow-2.3.1-foss-2020a-Python-3.8.2.eb|wc -l
0
```
This way you list all the missing deps, filter out the one you're building and then can just check whether the result is zero.

Reply:
Good idea! Added a solution based on this. Example:
```
$ ./findPythonDeps.py --ec TensorFlow-2.9.1-foss-2022a-CUDA-11.7.0.eb foo
You need to install all modules on which TensorFlow-2.9.1-foss-2022a-CUDA-11.7.0.eb depends first!
Missing:
	* pkgconf/1.8.0 (pkgconf-1.8.0.eb)
```